### PR TITLE
Removed libsodium; Made hashes more generic

### PR DESCRIPTION
**Pull request by mikelodder7**
Signed-off-by: Michael Lodder <redmike7@gmail.com>

### CODE REVIEW

File: libursa/src/signatures/ed25519.rs [hunk 1]
```diff
@@ -1,163 +1,58 @@
...
-        Ed25519Sha512(ed25519_sha2_512::Ed25519Sha512Impl::new())
+    fn new() -> Self {
+        Self
     }
     fn keypair(&self, option: Option<KeyPairOption>) -> Result<(PublicKey, PrivateKey), CryptoError> {
-        self.0.keypair(option)
+        let kp = match option {
+            Some(o) => match o {
+                KeyPairOption::UseSeed(s) => {
+                    let hash = sha2::Sha256::digest(s.as_slice());
```
Comment (lovesh) [inline libursa/src/signatures/ed25519.rs]: The caller of `keypair` will lose access to `s`. Should this function be zeroing `s`?

File: libursa/src/signatures/secp256k1.rs [hunk 2]
```diff
@@ -77,18 +79,19 @@ mod ecdsa_secp256k1sha256 {
...
         }
-        pub fn keypair(&self, option: Option<KeyPairOption>) -> Result<(PublicKey, PrivateKey), CryptoError> {
+        pub fn keypair<D>(&self, option: Option<KeyPairOption>) -> Result<(PublicKey, PrivateKey), CryptoError> where D: Digest<OutputSize = U32> {
             let sk = match option {
                     Some(o) => {
                         match o {
                             KeyPairOption::UseSeed(seed) => {
                                 let mut s = [0u8; PRIVATE_KEY_SIZE];
                                 let mut rng = ChaChaRng::from_seed(*array_ref!(seed.as_slice(), 0, 32));
                                 rng.fill_bytes(&mut s);
```
Comment (lovesh) [inline libursa/src/signatures/secp256k1.rs]: `s` will not be returned so the caller has no control to zero it? Should this code zero it or that is not a concern for now?

File: libursa/src/signatures/secp256k1.rs [hunk 3]
```diff
@@ -77,18 +79,19 @@ mod ecdsa_secp256k1sha256 {
...
-            EcdsaSecp256k1Sha256Impl(libsecp256k1::Secp256k1::new())
+        pub fn new() -> Self {
+            Self(libsecp256k1::Secp256k1::new())
         }
-        pub fn keypair(&self, option: Option<KeyPairOption>) -> Result<(PublicKey, PrivateKey), CryptoError> {
+        pub fn keypair<D>(&self, option: Option<KeyPairOption>) -> Result<(PublicKey, PrivateKey), CryptoError> where D: Digest<OutputSize = U32> {
             let sk = match option {
                     Some(o) => {
                         match o {
                             KeyPairOption::UseSeed(seed) => {
```
Comment (lovesh) [inline libursa/src/signatures/secp256k1.rs]: Who zeroes `seed`

File: libursa/src/signatures/secp256k1.rs [hunk 4]
```diff
@@ -97,21 +100,22 @@ mod ecdsa_secp256k1sha256 {
                         let mut rng = OsRng::new().map_err(|err| CryptoError::KeyGenError(format!("{}", err)))?;
                         let mut s = [0u8; PRIVATE_KEY_SIZE];
                         rng.fill_bytes(&mut s);
-                        libsecp256k1::key::SecretKey::from_slice(&s[..])?
+                        let s = D::digest(&s);
+                        libsecp256k1::key::SecretKey::from_slice(s.as_slice())?
```
Comment (lovesh) [inline libursa/src/signatures/secp256k1.rs]: Who zeroes this `s`?

File: libursa/src/signatures/secp256k1.rs [hunk 5]
```diff
@@ -97,21 +100,22 @@ mod ecdsa_secp256k1sha256 {
...
                 };
             let pk = libsecp256k1::key::PublicKey::from_secret_key(&self.0, &sk);
             Ok((PublicKey(pk.serialize().to_vec()), PrivateKey(sk[..].to_vec())))
         }
-        pub fn sign(&self, message: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, CryptoError> {
-            let h = digest(DigestAlgorithm::Sha2_256, message)?;
+        pub fn sign<D>(&self, message: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, CryptoError> where D: Digest<OutputSize = U32> {
+            let h = D::digest(message);
             let msg = libsecp256k1::Message::from_slice(h.as_slice())?;
             let s = libsecp256k1::key::SecretKey::from_slice(&sk[..])?;
```
Comment (lovesh) [inline libursa/src/signatures/secp256k1.rs]: `s` should be zeroed

File: libursa/src/signatures/ed25519.rs (hunk 1 above)
Comment (mikelodder7) [reply libursa/src/signatures/ed25519.rs]: ed25519-dalek implements this for you in its Drop trait.

File: libursa/src/signatures/secp256k1.rs (hunk 5 above)
Comment (mikelodder7) [reply libursa/src/signatures/secp256k1.rs]: `s` is zero'd in the Drop trait. See libursa/src/utils/macros.

**Pull request merged**